### PR TITLE
Improve the chatbot UX by scroll the user input message to the top after sending

### PR DESCRIPTION
**Pull request by FriedhelmWS**
### Description

https://github.com/user-attachments/assets/96f8e1e1-823d-49e2-abe8-7847ba2ffe82

### Issues Resolved
NA

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test.
- [ ] New functionality has user manual doc added.
- [ ] Commits are signed per the DCO using --signoff.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/tabs/chat/chat_page_content.tsx [hunk 1]
```diff
@@ -44,12 +44,27 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
   const chatActions = useChatActions();
   const registry = getIncontextInsightRegistry();
   const configSchema = getConfigSchema();
+  const latestInputRef = useRef<HTMLDivElement>(null);
+  const [messageSpacerHeight, setMessageSpacerHeight] = useState(0);
 
   useLayoutEffect(() => {
-    if (!props.chatScrollTopRef.current) {
+    if (latestInputRef.current) {
+      setMessageSpacerHeight(window.innerHeight - latestInputRef.current.clientHeight - 135);
```
Comment (yyfamazon) [inline public/tabs/chat/chat_page_content.tsx:52]: why you use 135?

Comment (FriedhelmWS) [reply public/tabs/chat/chat_page_content.tsx:52]: This is approximately the height of header + bottom input box + padding. Since this height is not responsive so it should be fine for us to hard-coded it here.

Comment (SuZhou-Joe) [reply public/tabs/chat/chat_page_content.tsx:52]: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/DEVELOPER_GUIDE.md#avoid-magic-numbersstrings, we should avoid magic number. Can we assign it to a variable and add some comments?

File: public/tabs/chat/chat_page_content.tsx [hunk 2]
```diff
@@ -300,3 +331,39 @@ const Suggestions: React.FC<SuggestionsProps> = (props) => {
...
+          minHeight: height,
+        }}
+      >
+        <MessageBubble loading showActionBar={false} />
+        {children}
+      </div>
+    );
+  }
+  return null;
+};
```
Comment (SuZhou-Joe) [inline public/tabs/chat/chat_page_content.tsx:369]: Let's move it to a new file.

File: release-notes/dashboards-assistant.release-notes-3.0.0.0-beta1.md [hunk 3]
```diff
@@ -15,6 +15,7 @@ Compatible with OpenSearch and OpenSearch Dashboards version 3.0.0-beta1
 - Change chatbot entry point to a single button ([#540](https://github.com/opensearch-project/dashboards-assistant/pull/540))
 - Support streaming output([#493](https://github.com/opensearch-project/dashboards-assistant/pull/493))
 - Update event names for t2v and feedback ([#543](https://github.com/opensearch-project/dashboards-assistant/pull/543))
+- Improve the chatbot UX by scroll the user input message to the top after sending ([#545](https://github.com/opensearch-project/dashboards-assistant/pull/545))
```
Comment (SuZhou-Joe) [inline release-notes/dashboards-assistant.release-notes-3.0.0.0-beta1.md:18]: Let's move this change to CHANGELOG.md instead of release notes.

File: public/tabs/chat/chat_page_content.tsx [hunk 4]
```diff
@@ -44,12 +44,27 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
...
   const configSchema = getConfigSchema();
+  const latestInputRef = useRef<HTMLDivElement>(null);
+  const [messageSpacerHeight, setMessageSpacerHeight] = useState(0);
 
   useLayoutEffect(() => {
-    if (!props.chatScrollTopRef.current) {
+    if (latestInputRef.current) {
+      setMessageSpacerHeight(window.innerHeight - latestInputRef.current.clientHeight - 135);
+
+      setTimeout(() => {
```
Comment (SuZhou-Joe) [inline public/tabs/chat/chat_page_content.tsx:54]: Nit: use `requestAnimationFrame`.

File: public/tabs/chat/chat_page_content.tsx [hunk 5]
```diff
@@ -44,12 +44,27 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
...
+        // Do the scrolling in the next loop to avoid page flashing
+        if (latestInputRef.current) {
+          latestInputRef.current.scrollIntoView({ behavior: 'smooth' });
+        }
+      }, 0);
+    }
+  }, [loading, latestInputRef, window.innerHeight]);
```
Comment (SuZhou-Joe) [inline public/tabs/chat/chat_page_content.tsx:61]: I am thinking if we should add `latestInputRef.current.clientHeight`.

Comment (FriedhelmWS) [reply public/tabs/chat/chat_page_content.tsx:61]: I am not sure that `latestInputRef.current`is possibly null so should we put it into dependency list?

File: public/tabs/chat/chat_page_content.tsx (hunk 4 above)
Comment (FriedhelmWS) [reply public/tabs/chat/chat_page_content.tsx:54]: Great suggestion, just checked that `requestAnimationFrame` is more suitable in this visual update case for a few reasons, thanks!

File: public/tabs/chat/chat_page_content.tsx (hunk 5 above)
Comment (SuZhou-Joe) [reply public/tabs/chat/chat_page_content.tsx:61]: Forget it, it should be fine.

File: public/tabs/chat/chat_page.tsx [hunk 6]
```diff
@@ -122,7 +122,7 @@ export const ChatPage: React.FC<ChatPageProps> = (props) => {
         <EuiSpacer size="xs" />
         <ChatInputControls
           loading={chatState.llmResponding}
-          disabled={messagesLoading || chatState.llmResponding}
+          disabled={messagesLoading || conversationsLoading || chatState.llmResponding}
```
Comment (SuZhou-Joe) [inline public/tabs/chat/chat_page.tsx:125]: Do not think it is good to disable ChatbotInput when conversations is loading. Perhaps we can update messagesLoading?

File: public/tabs/chat/chat_page_content.tsx [hunk 7]
```diff
@@ -44,12 +53,29 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
   const chatActions = useChatActions();
   const registry = getIncontextInsightRegistry();
   const configSchema = getConfigSchema();
+  const latestInputRef = useRef<HTMLDivElement>(null);
+  const [messageSpacerHeight, setMessageSpacerHeight] = useState(0);
 
   useLayoutEffect(() => {
-    if (!props.chatScrollTopRef.current) {
+    if (latestInputRef.current) {
+      setMessageSpacerHeight(
```
Comment (SuZhou-Joe) [inline public/tabs/chat/chat_page_content.tsx:61]: Shall we setMessageSpacerHeight when loading is true?

File: public/tabs/chat/chat_page_content.tsx [hunk 8]
```diff
@@ -163,8 +194,24 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
           (chatState.llmResponseType === LLMResponseType.TEXT ||
             (chatState.llmResponseType === LLMResponseType.STREAMING && !chatState.llmResponding));
 
+        // The latest llm output, after the llm has responded at least a message
+        const isStreaming =
+          isLatestOutput &&
+          chatState.llmResponseType === LLMResponseType.STREAMING &&
+          chatState.llmResponding;
```
Comment (SuZhou-Joe) [inline public/tabs/chat/chat_page_content.tsx:201]: Nit: isStreamingMessage?

File: public/tabs/chat/chat_page_content.tsx [hunk 9]
```diff
@@ -174,50 +221,42 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
...
+          )}
+      </LoadingPlaceholder>
       {chatState.llmError && (
-        <EuiEmptyPrompt
-          iconType="alert"
-          iconColor="danger"
-          title={<h2>Error from response</h2>}
-          body={chatState.llmError.message}
-        />
+        <div style={{ minHeight: messageSpacerHeight }}>
```
Comment (SuZhou-Joe) [inline public/tabs/chat/chat_page_content.tsx:252]: Add minHeight for error case is a edge case, can we add a unit test to cover it?

File: public/tabs/chat/loading_placeholder.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  height: number;
+  children?: ReactNode;
+}
+
+export const LoadingPlaceholder: React.FC<LoadingPlaceholderProps> = ({
+  loading,
+  height,
+  children,
+}: LoadingPlaceholderProps) => {
+  const [showLoadingPlaceholder, setShowLoadingPlaceholder] = useState(false);
```
Comment (SuZhou-Joe) [inline public/tabs/chat/loading_placeholder.tsx:20]: Seems using props.loading is good enough, do we need the state?

Comment (FriedhelmWS) [reply public/tabs/chat/loading_placeholder.tsx:20]: Nope, I used to do other thing in the useEffect but then removed, seems this useEffect is not necessary entirely and I forgot to also remove it, thanks for suggestion.

File: public/tabs/chat/chat_page_content.tsx (hunk 7 above)
Comment (FriedhelmWS) [reply public/tabs/chat/chat_page_content.tsx:61]: `latestInputRef` only attach to a element when loading is true, but good to also add a double check here for clarification anyway.

File: public/tabs/chat/chat_page_content.tsx (hunk 9 above)
Comment (FriedhelmWS) [reply public/tabs/chat/chat_page_content.tsx:252]: Done, add a test case to cover this line

File: public/tabs/chat/chat_page.tsx (hunk 6 above)
Comment (FriedhelmWS) [reply public/tabs/chat/chat_page.tsx]: Discussed offline, will revert this fix and leave it for a later PR

**Review by SuZhou-Joe — APPROVED** [review]

File: public/tabs/chat/chat_page_content.tsx [hunk 11]
```diff
@@ -44,12 +53,29 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
...
   const registry = getIncontextInsightRegistry();
   const configSchema = getConfigSchema();
+  const latestInputRef = useRef<HTMLDivElement>(null);
+  const [messageSpacerHeight, setMessageSpacerHeight] = useState(0);
 
   useLayoutEffect(() => {
-    if (!props.chatScrollTopRef.current) {
+    if (loading && latestInputRef.current) {
+      setMessageSpacerHeight(
+        window.innerHeight - latestInputRef.current.clientHeight - CHAT_PAGE_CONTENT_PADDING_HEIGHT
```
Comment (wanglam) [inline public/tabs/chat/chat_page_content.tsx:62]: Do we need to consider `latestInputRef.current.clientHeight` greater than `window.innerHeight` here?

Comment (FriedhelmWS) [reply public/tabs/chat/chat_page_content.tsx:62]: If the height value becomes negative, then setting it to minHeight would be equivalent to setting minHeight to 0, resulting in no spacing being provided, which is expected. But do you think it is necessary to add a conditional check here?

File: public/tabs/chat/chat_page_content.tsx [hunk 12]
```diff
@@ -174,50 +221,42 @@ export const ChatPageContent: React.FC<ChatPageContentProps> = React.memo((props
...
-                  chatState.llmResponseType === LLMResponseType.STREAMING &&
-                  chatState.llmResponding
-                }
-              />
+              <MessageContent message={message} loading={isStreamingMessage} />
             </MessageBubble>
-            {showSuggestions && <Suggestions message={message} inputDisabled={loading} />}
             <EuiSpacer />
-          </React.Fragment>
+            {showSuggestions && !isStreamingMessage && (
```
[the remark this answers is not in the document]
Comment (FriedhelmWS) [reply public/tabs/chat/chat_page_content.tsx:227]: Nah, it already been fix in a previous PR, will remove the fix here

**Review by wanglam — APPROVED** [review]

File: public/tabs/chat/chat_page_content.tsx (hunk 11 above)
Comment (wanglam) [reply public/tabs/chat/chat_page_content.tsx:62]: I'm not sure if we need to use `latestInputRef.current.clientHeight - window.innerHeight` if input height greater than `window.innerHeight`. Because I think the current response should still at the top of chat page content. I'm OK to leave it as current implementation. Let's fix this in the future PR.
